File: server/src/fft-processor.ts
// ============================================================
// node-sdr — Server-Side FFT Processor
// ============================================================
// Converts raw IQ data from rtl_sdr into FFT magnitude (dB) arrays.
// Uses fft.js (indutny) — radix-4, fastest pure JS FFT for typical
// sizes (≤8192). For large FFT sizes (≥32768) the number[] overhead
// becomes significant, but radix-4 algorithmic advantage wins at
// the common 2048–4096 range used in production.
//
// IQ data from rtl_sdr: unsigned 8-bit interleaved [I0,Q0,I1,Q1,...]
// Output: Float32Array of dB magnitude values per FFT bin.
// ============================================================

// @ts-ignore — fft.js has no type declarations
import FFT from 'fft.js';
import { logger } from './logger.js';

export interface FftProcessorOptions {
  /** FFT size (power of 2) */
  fftSize: number;
  /** Sample rate for this dongle */
  sampleRate: number;
  /** Window function to apply */
  window?: 'blackman-harris' | 'hann' | 'hamming' | 'none';
  /** Averaging factor (0 = no averaging, 0.9 = heavy smoothing) */
  averaging: number;
  /** Target FFT output frame rate (fps). 0 = unlimited. Default 30. */
  targetFps?: number;
}

export class FftProcessor {
  private fft: any;
  private fftSize: number;
  private windowFunc: Float32Array;
  private averagedMagnitudes: Float32Array | null = null;
  private readonly averaging: number;

  // Reusable buffers for fft.js (requires number[] — radix-4 is faster at N≤8192)
  private complexInput: number[];
  private complexOutput: number[];

  // Pre-allocated magnitude scratch buffer — reused every processOneFrame call
  private magnitudesBuffer: Float32Array;

  // Ring buffer accumulator — pre-allocated, no Buffer.concat on every chunk
  private iqRingBuf: Buffer;
  private iqRingFill = 0;
  private readonly samplesNeeded: number;

  // FFT normalization: 10 * log10(N²) + window coherent gain correction
  private normalizationDb: number;

  // Rate limiting
  private targetFps: number;
  private minFrameInterval: number;
  private lastEmitTime = 0;
  // Pre-allocated pending frame buffer — avoids new Float32Array(fftSize) at 30fps.
  // framesAccumulated === 0 means the buffer is empty/unused for the current interval.
  private pendingFrame: Float32Array;
  private framesAccumulated = 0;

  constructor(private options: FftProcessorOptions) {
    this.fftSize = options.fftSize;
    this.averaging = options.averaging;
    this.fft = new FFT(this.fftSize);
    this.complexInput = this.fft.createComplexArray();
    this.complexOutput = this.fft.createComplexArray();
    this.windowFunc = this.createWindow(options.window ?? 'blackman-harris');
    this.samplesNeeded = this.fftSize * 2;
    this.normalizationDb = this.computeNormalization();
    this.targetFps = options.targetFps ?? 30;
    this.minFrameInterval = this.targetFps > 0 ? 1000 / this.targetFps : 0;

    // Pre-allocate magnitude scratch buffer, ring buffer (4 frames of headroom),
    // and pending frame buffer (one per emit interval — no alloc at 30fps).
    this.magnitudesBuffer = new Float32Array(this.fftSize);
    this.pendingFrame = new Float32Array(this.fftSize);
    this.iqRingBuf = Buffer.allocUnsafe(this.samplesNeeded * 4);

    logger.debug(
      { fftSize: this.fftSize, window: options.window ?? 'blackman-harris', normalizationDb: this.normalizationDb.toFixed(1), targetFps: this.targetFps },
      'FFT processor initialized',
    );
  }

  processIqData(rawData: Buffer): Float32Array[] {
    // Append into ring buffer — no alloc, no copy of existing data
    if (this.iqRingFill + rawData.length > this.iqRingBuf.length) {
      // Ring buffer too small (shouldn't happen with 4× headroom) — grow it
      const newBuf = Buffer.allocUnsafe((this.iqRingFill + rawData.length) * 2);
      this.iqRingBuf.copy(newBuf, 0, 0, this.iqRingFill);
      this.iqRingBuf = newBuf;
    }
    rawData.copy(this.iqRingBuf, this.iqRingFill);
    this.iqRingFill += rawData.length;

    const emitted: Float32Array[] = [];
    const now = Date.now(); // hoist out of loop — constant for this call

    let consumed = 0;
    while (this.iqRingFill - consumed >= this.samplesNeeded) {
      const frame = this.processOneFrame(this.iqRingBuf.subarray(consumed, consumed + this.samplesNeeded));
      consumed += this.samplesNeeded;

      if (this.minFrameInterval <= 0) {
        emitted.push(frame);
        continue;
      }

      if (this.framesAccumulated === 0) {
        // First frame of interval: copy into pre-allocated pendingFrame — no alloc
        this.pendingFrame.set(frame);
        this.framesAccumulated = 1;
      } else {
        // Incremental mean: avg += (new - avg) / count
        this.framesAccumulated++;
        const n = this.framesAccumulated;
        const pf = this.pendingFrame;
        for (let i = 0; i < frame.length; i++) {
          pf[i] += (frame[i] - pf[i]) / n;
        }
      }

      if (now - this.lastEmitTime >= this.minFrameInterval) {
        emitted.push(this.pendingFrame);
        this.framesAccumulated = 0;
        this.lastEmitTime = now;
      }
    }

    // Compact ring buffer: move remainder to front
    if (consumed > 0) {
      if (consumed < this.iqRingFill) {
        this.iqRingBuf.copyWithin(0, consumed, this.iqRingFill);
      }
      this.iqRingFill -= consumed;
    }

    return emitted;
  }

  private processOneFrame(rawIq: Buffer): Float32Array {
    // Convert uint8 IQ → float [-1, +1] and apply window into fft.js input array
    for (let i = 0; i < this.fftSize; i++) {
      const w = this.windowFunc[i];
      this.complexInput[i * 2]     = ((rawIq[i * 2]     - 127.5) / 127.5) * w;
      this.complexInput[i * 2 + 1] = ((rawIq[i * 2 + 1] - 127.5) / 127.5) * w;
    }

    this.fft.transform(this.complexOutput, this.complexInput);

    // Compute magnitude in dB into reusable scratch buffer — no allocation
    const magnitudes = this.magnitudesBuffer;
    const fftSize = this.fftSize;
    const half = fftSize >> 1;

    for (let i = 0; i < fftSize; i++) {
      // FFT-shift: remap so DC is in the center
      const srcIdx = (i + half) & (fftSize - 1);
      const re = this.complexOutput[srcIdx * 2];
      const im = this.complexOutput[srcIdx * 2 + 1];
      magnitudes[i] = 10 * Math.log10(Math.max(re * re + im * im, 1e-20)) - this.normalizationDb;
    }

    // Apply exponential averaging
    if (this.averaging > 0) {
      if (!this.averagedMagnitudes) {
        this.averagedMagnitudes = new Float32Array(magnitudes);
      } else {
        const a = this.averaging;
        const b = 1 - a;
        for (let i = 0; i < fftSize; i++) {
          this.averagedMagnitudes[i] = a * this.averagedMagnitudes[i] + b * magnitudes[i];
        }
      }
      return this.averagedMagnitudes;
    }

    return magnitudes;
  }

  /**
   * Create a window function array
   */
  private createWindow(type: string): Float32Array {
    const N = this.fftSize;
    const w = new Float32Array(N);

    switch (type) {
      case 'blackman-harris': {
        const a0 = 0.35875, a1 = 0.48829, a2 = 0.14128, a3 = 0.01168;
        for (let i = 0; i < N; i++) {
          const x = (2 * Math.PI * i) / (N - 1);
          w[i] = a0 - a1 * Math.cos(x) + a2 * Math.cos(2 * x) - a3 * Math.cos(3 * x);
        }
        break;
      }
      case 'hann': {
        for (let i = 0; i < N; i++) {
          w[i] = 0.5 * (1 - Math.cos((2 * Math.PI * i) / (N - 1)));
        }
        break;
      }
      case 'hamming': {
        for (let i = 0; i < N; i++) {
          w[i] = 0.54 - 0.46 * Math.cos((2 * Math.PI * i) / (N - 1));
        }
        break;
      }
      default: {
        w.fill(1);
        break;
      }
    }

    return w;
  }

  /**
   * Compute FFT normalization in dB.
   * Accounts for FFT size (N²) and window function coherent gain.
   * The coherent gain of a window is sum(w[i]) / N.
   */
  private computeNormalization(): number {
    // FFT size normalization: 10 * log10(N²) = 20 * log10(N)
    const fftNorm = 20 * Math.log10(this.fftSize);

    // Window coherent gain: sum(w) / N
    let windowSum = 0;
    for (let i = 0; i < this.fftSize; i++) {
      windowSum += this.windowFunc[i];
    }
    const coherentGain = windowSum / this.fftSize;
    // Window correction in dB (for power): 20 * log10(coherentGain)
    const windowCorrectionDb = 20 * Math.log10(coherentGain);

    // Total normalization: subtract this from raw 10*log10(|X|²)
    return fftNorm + windowCorrectionDb;
  }

  /**
   * Reset internal buffers (call when switching profiles)
   */
  reset(): void {
    this.iqRingFill = 0;
    this.averagedMagnitudes = null;
    this.framesAccumulated = 0;
    this.lastEmitTime = 0;
  }

  resize(newFftSize: number): void {
    this.fftSize = newFftSize;
    this.fft = new FFT(newFftSize);
    this.complexInput = this.fft.createComplexArray();
    this.complexOutput = this.fft.createComplexArray();
    this.windowFunc = this.createWindow(this.options.window ?? 'blackman-harris');
    this.averagedMagnitudes = null;
    this.normalizationDb = this.computeNormalization();
    (this as any).samplesNeeded = newFftSize * 2;
    this.magnitudesBuffer = new Float32Array(newFftSize);
    this.pendingFrame = new Float32Array(newFftSize);
    this.iqRingBuf = Buffer.allocUnsafe(newFftSize * 2 * 4);
    this.iqRingFill = 0;
    this.framesAccumulated = 0;
  }
}

/**
 * Extract a sub-band of IQ data centered around a frequency offset.
 * Used to send per-user IQ data for client-side demodulation.
 *
 * @param fftOutput - Complex FFT output (interleaved re/im)
 * @param fftSize - FFT size
 * @param sampleRate - Full sample rate
 * @param centerOffsetHz - Desired center frequency offset from DC
 * @param bandwidthHz - Desired bandwidth
 * @returns Int16Array of interleaved I/Q sub-band samples
 */
export function extractIqSubBand(
  fftOutput: number[],
  fftSize: number,
  sampleRate: number,
  centerOffsetHz: number,
  bandwidthHz: number,
): Int16Array {
  const binWidth = sampleRate / fftSize;
  const centerBin = Math.round(centerOffsetHz / binWidth) + fftSize / 2;
  const halfBins = Math.round(bandwidthHz / binWidth / 2);

  const startBin = Math.max(0, centerBin - halfBins);
  const endBin = Math.min(fftSize - 1, centerBin + halfBins);
  const numBins = endBin - startBin + 1;

  // IFFT the sub-band to get time-domain IQ samples
  // For now, just extract the frequency-domain bins
  // The client will do the IFFT
  const subBand = new Int16Array(numBins * 2);

  for (let i = 0; i < numBins; i++) {
    const srcBin = startBin + i;
    const re = fftOutput[srcBin * 2];
    const im = fftOutput[srcBin * 2 + 1];
    // Scale to Int16 range
    subBand[i * 2] = Math.max(-32768, Math.min(32767, Math.round(re * 32767)));
    subBand[i * 2 + 1] = Math.max(-32768, Math.min(32767, Math.round(im * 32767)));
  }

  return subBand;
}

File: serverng/internal/dsp/fft_processor.go
package dsp

import (
	"errors"
	"log/slog"
	"math"
	"time"
	"unsafe"
)

// GPUFFTBackend is the interface that a GPU FFT provider must satisfy.
// The gpu.FFTContext type satisfies this interface.
type GPUFFTBackend interface {
	// Process runs a C2C FFT on fftSize*2 uint8 IQ bytes.
	// Returns fftSize float32 values: 10*log10(power) for each bin in natural FFT order.
	// Caller applies FFT-shift and subtracts normDbVal.
	// Returns a non-nil error on GPU failure — caller must fall back to CPU.
	Process(iq []byte) ([]float32, error)
	// Close releases the GPU resources.
	Close()
}

// fastLog10Power computes an approximation of 10*log10(x) using the IEEE 754
// bit representation of x. It extracts the binary exponent directly and uses
// a 3-term minimax polynomial over the mantissa for the fractional log2 part,
// then multiplies by log10(2). Max error is < 0.01 dB — imperceptible on a
// waterfall display. x must be positive.
//
// Derivation: log10(x) = log2(x) * log10(2)
//   log2(x) = exponent + log2(mantissa),  mantissa in [1,2)
//   log2(m) ≈ -1.0 + 2.0*m - 0.5*m*m  (minimax on [1,2], max err ~0.006)
func fastLog10Power(x float32) float32 {
	// Reinterpret bits to access the raw IEEE 754 float32 word.
	bits32 := *(*uint32)(unsafe.Pointer(&x))

	// Extract unbiased binary exponent: (bits >> 23) - 127
	exp := int32(bits32>>23) - 127

	// Build a float32 with the same mantissa but exponent = 0 (value in [1,2)).
	mantissaBits := (bits32 & 0x7FFFFF) | 0x3F800000
	m := *(*float32)(unsafe.Pointer(&mantissaBits))

	// 3-term minimax polynomial for log2(m), m in [1,2):
	//   log2(m) ≈ -1.0 + 2.0*m - 0.5*m*m  (max error ~0.006 bits)
	log2m := -1.0 + 2.0*m - 0.5*m*m

	// 10*log10(x) = log2(x) * 10*log10(2)
	const tenLog10Of2 = float32(3.01029995663981) // 10 * log10(2)
	return (float32(exp) + log2m) * tenLog10Of2
}

// FftProcessorConfig configures the FFT processor pipeline.
type FftProcessorConfig struct {
	FftSize    int     // power of 2
	SampleRate int     // dongle sample rate
	Window     string  // "blackman-harris", "hann", "hamming"
	Averaging  float32 // 0 = no averaging, 0.9 = heavy smoothing
	TargetFps  int     // output frame rate cap (0 = unlimited). Typical: 30.
}

// FftProcessor accumulates raw uint8 IQ data, applies windowing + FFT,
// computes magnitude in dB with averaging, and rate-caps output frames.
type FftProcessor struct {
	fftSize    int
	sampleRate int
	averaging  float32
	targetFps  int
	windowName string

	// DSP core
	fft       *FFT
	window    []float32
	normDbVal float32

	// GPU FFT backend (optional — nil = CPU only)
	gpuFFT       GPUFFTBackend
	gpuPaused    bool  // when true, GPU path is skipped (admin disabled)
	gpuFrames    int64 // count of frames processed by GPU

	// Pre-allocated work buffers (zero alloc in hot path)
	complexBuf []float32 // interleaved complex input/output, length 2*fftSize
	magBuf     []float32 // magnitude output, length fftSize
	avgBuf     []float32 // exponential averaging state, length fftSize (nil if averaging==0)
	avgInit    bool       // whether avgBuf has been initialized with first frame

	// Ring buffer accumulator
	ringBuf     []byte
	ringFill    int
	samplesNeed int // fftSize * 2 bytes

	// Rate cap state
	minInterval time.Duration
	lastEmit    time.Time
	pendingBuf  []float32 // accumulated frame for rate cap, length fftSize
	pendingN    int       // number of frames accumulated into pendingBuf
}

// NewFftProcessor creates a new FFT processor with fully pre-allocated buffers.
func NewFftProcessor(cfg FftProcessorConfig) (*FftProcessor, error) {
	if cfg.FftSize < 4 || cfg.FftSize&(cfg.FftSize-1) != 0 {
		return nil, errors.New("fft_processor: fftSize must be a power of 2 and >= 4")
	}
	if cfg.Window == "" {
		cfg.Window = "blackman-harris"
	}
	if cfg.Averaging < 0 {
		cfg.Averaging = 0
	}
	if cfg.Averaging > 1 {
		cfg.Averaging = 1
	}

	fft, err := NewFFT(cfg.FftSize)
	if err != nil {
		return nil, err
	}

	win, err := NewWindow(cfg.Window, cfg.FftSize)
	if err != nil {
		return nil, err
	}

	p := &FftProcessor{
		fftSize:    cfg.FftSize,
		sampleRate: cfg.SampleRate,
		averaging:  cfg.Averaging,
		targetFps:  cfg.TargetFps,
		windowName: cfg.Window,
		fft:        fft,
		window:     win,
	}

	p.normDbVal = p.computeNormalization()
	p.samplesNeed = cfg.FftSize * 2

	// Pre-allocate all buffers
	p.complexBuf = make([]float32, 2*cfg.FftSize)
	p.magBuf = make([]float32, cfg.FftSize)
	p.ringBuf = make([]byte, cfg.FftSize*2*4) // 4 frames of headroom
	p.ringFill = 0
	p.pendingBuf = make([]float32, cfg.FftSize)
	p.pendingN = 0

	if cfg.Averaging > 0 {
		p.avgBuf = make([]float32, cfg.FftSize)
		p.avgInit = false
	}

	// Rate cap
	if cfg.TargetFps > 0 {
		p.minInterval = time.Second / time.Duration(cfg.TargetFps)
	}

	return p, nil
}

// SetGPUBackend attaches a GPU FFT backend to this processor.
// Calling with nil removes the GPU backend and falls back to CPU.
// Must not be called concurrently with ProcessIqData.
func (p *FftProcessor) SetGPUBackend(g GPUFFTBackend) {
	if p.gpuFFT != nil {
		p.gpuFFT.Close()
	}
	p.gpuFFT = g
	if g != nil {
		slog.Info("fft_processor: GPU FFT backend attached",
			"fft_size", p.fftSize)
	} else {
		slog.Info("fft_processor: GPU FFT backend removed, falling back to CPU")
	}
}

// SetGPUPaused pauses/resumes the GPU FFT path without tearing down the backend.
func (p *FftProcessor) SetGPUPaused(paused bool) {
	p.gpuPaused = paused
}

// GPUFrames returns the number of FFT frames processed by the GPU backend.
func (p *FftProcessor) GPUFrames() int64 {
	return p.gpuFrames
}

// ProcessIqData accepts raw uint8 IQ data (interleaved I,Q,I,Q...)
// and returns zero or more FFT magnitude frames ([]float32 in dB, length = FftSize).
// May return 0 frames if rate cap hasn't elapsed or insufficient data accumulated.
// The returned slices are owned by the caller (copied from internal buffers).
func (p *FftProcessor) ProcessIqData(data []byte) [][]float32 {
	// Append into ring buffer
	needed := p.ringFill + len(data)
	if needed > len(p.ringBuf) {
		// Grow ring buffer (rare path)
		newBuf := make([]byte, needed*2)
		copy(newBuf, p.ringBuf[:p.ringFill])
		p.ringBuf = newBuf
	}
	copy(p.ringBuf[p.ringFill:], data)
	p.ringFill += len(data)

	var emitted [][]float32
	now := time.Now()

	consumed := 0
	for p.ringFill-consumed >= p.samplesNeed {
		p.processOneFrame(p.ringBuf[consumed : consumed+p.samplesNeed])
		consumed += p.samplesNeed

		if p.minInterval <= 0 {
			// No rate cap — emit every frame
			frame := make([]float32, p.fftSize)
			copy(frame, p.magBuf)
			emitted = append(emitted, frame)
			continue
		}

		// Rate-capped: accumulate into pending frame
		if p.pendingN == 0 {
			copy(p.pendingBuf, p.magBuf)
			p.pendingN = 1
		} else {
			p.pendingN++
			n := float32(p.pendingN)
			for i := 0; i < p.fftSize; i++ {
				p.pendingBuf[i] += (p.magBuf[i] - p.pendingBuf[i]) / n
			}
		}

		if now.Sub(p.lastEmit) >= p.minInterval {
			frame := make([]float32, p.fftSize)
			copy(frame, p.pendingBuf)
			emitted = append(emitted, frame)
			p.pendingN = 0
			p.lastEmit = now
		}
	}

	// Compact ring buffer: move remainder to front
	if consumed > 0 {
		if consumed < p.ringFill {
			copy(p.ringBuf, p.ringBuf[consumed:p.ringFill])
		}
		p.ringFill -= consumed
	}

	return emitted
}

// processOneFrame converts one FFT-sized chunk of uint8 IQ into magnitude dB.
// Result is written into p.magBuf. Zero allocations on CPU path.
// If a GPU backend is attached, the GPU path is tried first with CPU fallback.
func (p *FftProcessor) processOneFrame(rawIq []byte) {
	// ── GPU path ──────────────────────────────────────────────────────────────
	if p.gpuFFT != nil && !p.gpuPaused {
		if gpuOut, err := p.gpuFFT.Process(rawIq); err == nil {
			// GPU returns fftSize bins in natural FFT order (DC at bin 0).
			// Apply FFT-shift (DC → center) and normalization — same as CPU path.
			fftSize := p.fftSize
			half := fftSize >> 1
			normDb := p.normDbVal
			mag := p.magBuf

			for i := 0; i < fftSize; i++ {
				srcIdx := (i + half) & (fftSize - 1)
				mag[i] = gpuOut[srcIdx] - normDb
			}

			// Apply exponential averaging (same as CPU path)
			if p.averaging > 0 {
				if !p.avgInit {
					copy(p.avgBuf, mag)
					p.avgInit = true
				} else {
					a := p.averaging
					b := 1 - a
					avg := p.avgBuf
					for i := 0; i < fftSize; i++ {
						avg[i] = a*avg[i] + b*mag[i]
					}
				}
				copy(mag, p.avgBuf)
			}
			p.gpuFrames++
			return
		}
		// GPU failed — fall through to CPU path (do not disable GPU; transient errors are common)
	}

	// ── CPU path ──────────────────────────────────────────────────────────────
	fftSize := p.fftSize
	win := p.window
	buf := p.complexBuf

	// Convert uint8 IQ -> float32 [-1,+1] with windowing in one pass
	for i := 0; i < fftSize; i++ {
		w := win[i]
		buf[i*2] = (float32(rawIq[i*2]) - 127.5) / 127.5 * w
		buf[i*2+1] = (float32(rawIq[i*2+1]) - 127.5) / 127.5 * w
	}

	// In-place FFT
	p.fft.Transform(buf)

	// Compute magnitude in dB with FFT-shift (DC in center)
	half := fftSize >> 1
	normDb := p.normDbVal
	mag := p.magBuf

	for i := 0; i < fftSize; i++ {
		srcIdx := (i + half) & (fftSize - 1)
		re := buf[srcIdx*2]
		im := buf[srcIdx*2+1]
		power := re*re + im*im
		if power < 1e-20 {
			power = 1e-20
		}
		mag[i] = fastLog10Power(power) - normDb
	}

	// Exponential averaging
	if p.averaging > 0 {
		if !p.avgInit {
			copy(p.avgBuf, mag)
			p.avgInit = true
		} else {
			a := p.averaging
			b := 1 - a
			avg := p.avgBuf
			for i := 0; i < fftSize; i++ {
				avg[i] = a*avg[i] + b*mag[i]
			}
		}
		// Output the averaged magnitudes
		copy(mag, p.avgBuf)
	}
}

// Resize changes FFT size at runtime (used on profile switch).
func (p *FftProcessor) Resize(newSize int) error {
	if newSize < 4 || newSize&(newSize-1) != 0 {
		return errors.New("fft_processor: newSize must be a power of 2 and >= 4")
	}

	fft, err := NewFFT(newSize)
	if err != nil {
		return err
	}

	win, err := NewWindow(p.windowName, newSize)
	if err != nil {
		return err
	}

	p.fftSize = newSize
	p.fft = fft
	p.window = win
	p.samplesNeed = newSize * 2
	p.normDbVal = p.computeNormalization()

	// Re-allocate buffers
	p.complexBuf = make([]float32, 2*newSize)
	p.magBuf = make([]float32, newSize)
	p.pendingBuf = make([]float32, newSize)
	p.pendingN = 0
	p.ringBuf = make([]byte, newSize*2*4)
	p.ringFill = 0

	if p.averaging > 0 {
		p.avgBuf = make([]float32, newSize)
		p.avgInit = false
	} else {
		p.avgBuf = nil
		p.avgInit = false
	}

	return nil
}

// Reset clears averaging state and ring buffer.
func (p *FftProcessor) Reset() {
	p.ringFill = 0
	p.pendingN = 0
	p.lastEmit = time.Time{}
	p.avgInit = false
	if p.avgBuf != nil {
		for i := range p.avgBuf {
			p.avgBuf[i] = 0
		}
	}
}

// computeNormalization returns 20*log10(fftSize) + 20*log10(coherentGain)
// where coherentGain = sum(window) / fftSize.
func (p *FftProcessor) computeNormalization() float32 {
	fftNorm := 20.0 * math.Log10(float64(p.fftSize))

	var windowSum float64
	for _, v := range p.window {
		windowSum += float64(v)
	}
	coherentGain := windowSum / float64(p.fftSize)
	windowCorrectionDb := 20.0 * math.Log10(coherentGain)

	return float32(fftNorm + windowCorrectionDb)
}
